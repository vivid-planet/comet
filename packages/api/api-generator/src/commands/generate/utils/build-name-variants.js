"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildNameVariants = buildNameVariants;
var pluralize_1 = require("pluralize");
function classNameToInstanceName(className) {
    return className[0].toLocaleLowerCase() + className.slice(1);
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function buildNameVariants(metadata) {
    var classNameSingular = metadata.className;
    var classNamePlural = (0, pluralize_1.plural)(metadata.className);
    var instanceNameSingular = classNameToInstanceName(classNameSingular);
    var instanceNamePlural = classNameToInstanceName(classNamePlural);
    var fileNameSingular = instanceNameSingular.replace(/[A-Z]/g, function (i) { return "-".concat(i.toLocaleLowerCase()); });
    var fileNamePlural = instanceNamePlural.replace(/[A-Z]/g, function (i) { return "-".concat(i.toLocaleLowerCase()); });
    return {
        classNameSingular: classNameSingular,
        classNamePlural: classNamePlural,
        instanceNameSingular: instanceNameSingular,
        instanceNamePlural: instanceNamePlural,
        fileNameSingular: fileNameSingular,
        fileNamePlural: fileNamePlural,
    };
}
