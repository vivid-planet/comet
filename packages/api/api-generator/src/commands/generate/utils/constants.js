"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.integerTypes = void 0;
exports.integerTypes = ["int", "integer", "tinyint", "smallint", "mediumint", "bigint", "int2", "int4", "int8", "serial"];
