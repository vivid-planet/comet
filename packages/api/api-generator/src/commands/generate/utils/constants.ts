export const integerTypes = ["int", "integer", "tinyint", "smallint", "mediumint", "bigint", "int2", "int4", "int8", "serial"];
