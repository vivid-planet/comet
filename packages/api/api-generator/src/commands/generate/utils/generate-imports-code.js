"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateImportsCode = generateImportsCode;
// generate imports code and filter duplicates
function generateImportsCode(imports) {
    var importsNameToPath = new Map(); // name -> importPath
    var filteredImports = imports.filter(function (imp) {
        if (importsNameToPath.has(imp.name)) {
            if (importsNameToPath.get(imp.name) !== imp.importPath) {
                throw new Error("Duplicate import name ".concat(imp.name));
            }
            else {
                // duplicate import, skip
                return false;
            }
        }
        importsNameToPath.set(imp.name, imp.importPath);
        return true;
    });
    var importsPathToName = {};
    for (var _i = 0, filteredImports_1 = filteredImports; _i < filteredImports_1.length; _i++) {
        var imp = filteredImports_1[_i];
        if (!importsPathToName[imp.importPath]) {
            importsPathToName[imp.importPath] = [];
        }
        importsPathToName[imp.importPath].push(imp.name);
    }
    var importsString = "";
    for (var _a = 0, _b = Object.entries(importsPathToName); _a < _b.length; _a++) {
        var _c = _b[_a], importPath = _c[0], importNames = _c[1];
        importsString += "import { ".concat(importNames.sort().join(", "), " } from \"").concat(importPath, "\";\n");
    }
    return importsString;
}
