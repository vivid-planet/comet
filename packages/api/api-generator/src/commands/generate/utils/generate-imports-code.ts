export type Imports = Array<{
    name: string;
    importPath: string;
}>;

// generate imports code and filter duplicates
export function generateImportsCode(imports: Imports): string {
    const importsNameToPath: Map<string, string> = new Map(); // name -> importPath
    const filteredImports = imports.filter((imp) => {
        if (importsNameToPath.has(imp.name)) {
            if (importsNameToPath.get(imp.name) !== imp.importPath) {
                throw new Error(`Duplicate import name ${imp.name}`);
            } else {
                // duplicate import, skip
                return false;
            }
        }
        importsNameToPath.set(imp.name, imp.importPath);
        return true;
    });

    const importsPathToName: Record<string, string[]> = {};
    for (const imp of filteredImports) {
        if (!importsPathToName[imp.importPath]) {
            importsPathToName[imp.importPath] = [];
        }
        importsPathToName[imp.importPath].push(imp.name);
    }

    let importsString = "";
    for (const [importPath, importNames] of Object.entries(importsPathToName)) {
        importsString += `import { ${importNames.sort().join(", ")} } from "${importPath}";\n`;
    }
    return importsString;
}
