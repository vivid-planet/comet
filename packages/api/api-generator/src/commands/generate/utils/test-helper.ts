import * as process from "node:process";

import { type Permission } from "@comet/cms-api";
import { format, type Options, resolveConfig } from "prettier";
import { Project, type SourceFile } from "ts-morph";

import { type GeneratedFile } from "./write-generated-files.js";

let options: Options | null;
export async function formatSource(sourceCode: string): Promise<string> {
    if (!options) options = await resolveConfig(process.cwd());

    return format(sourceCode, {
        ...options,
        filepath: "test.ts",
    });
}

export async function formatGeneratedFiles(files: GeneratedFile[]): Promise<GeneratedFile[]> {
    return Promise.all(
        files.map(async (file) => {
            return {
                ...file,
                content: await formatSource(file.content),
            };
        }),
    );
}

export function parseSource(source: string): SourceFile {
    const project = new Project({
        tsConfigFilePath: "./tsconfig.json",
        skipAddingFilesFromTsConfig: true,
    });
    return project.createSourceFile("test.ts", source);
}

export const testPermission = "crud" as Permission;
