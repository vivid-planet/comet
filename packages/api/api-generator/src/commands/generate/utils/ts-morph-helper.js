"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.morphTsProperty = morphTsProperty;
exports.findImportPath = findImportPath;
exports.findEnumName = findEnumName;
exports.findEnumImportPath = findEnumImportPath;
exports.findValidatorImportPath = findValidatorImportPath;
exports.findBlockName = findBlockName;
exports.findBlockImportPath = findBlockImportPath;
exports.findInputClassImportPath = findInputClassImportPath;
exports.getFieldDecoratorClassName = getFieldDecoratorClassName;
var path = require("path");
var ts_morph_1 = require("ts-morph");
/* eslint-disable @typescript-eslint/no-explicit-any */
var project = new ts_morph_1.Project({
    tsConfigFilePath: "tsconfig.json",
    skipAddingFilesFromTsConfig: true,
});
function morphTsSource(metadata) {
    var tsSource = project.getSourceFile(metadata.path);
    if (!tsSource)
        tsSource = project.addSourceFileAtPath(metadata.path);
    return tsSource;
}
function morphTsClass(metadata) {
    var tsSource = morphTsSource(metadata);
    var tsClass = tsSource.getClass(metadata.className);
    if (!tsClass)
        throw new Error("Class ".concat(metadata.className, " not found in ").concat(metadata.path));
    return tsClass;
}
function morphTsProperty(name, metadata) {
    var currentClass = morphTsClass(metadata);
    while (currentClass) {
        var prop = currentClass.getProperty(name);
        if (prop)
            return prop;
        currentClass = currentClass.getBaseClass();
    }
    throw new Error("Property ".concat(name, " not found in ").concat(metadata.className));
}
function findImportPath(importName, targetDirectory, metadata) {
    var _a;
    var tsSource = morphTsSource(metadata);
    for (var _i = 0, _b = tsSource.getImportDeclarations(); _i < _b.length; _i++) {
        var tsImport = _b[_i];
        for (var _c = 0, _d = tsImport.getNamedImports(); _c < _d.length; _c++) {
            var namedImport = _d[_c];
            if (((_a = namedImport.getAliasNode()) === null || _a === void 0 ? void 0 : _a.getText()) == importName) {
                throw new Error("Aliased enum import is not supported");
            }
            if (namedImport.getNameNode().getText() == importName) {
                var importPath = tsImport.getModuleSpecifierValue();
                var importSource = tsImport.getModuleSpecifierSourceFile();
                if (!importSource) {
                    throw new Error("Can't resolve import ".concat(importPath));
                }
                var exportedDeclarations = importSource.getExportedDeclarations().get(importName);
                if (!exportedDeclarations)
                    throw new Error("Can't get exported declaration for ".concat(importName));
                var exportedDeclaration = exportedDeclarations[0];
                if (importPath.startsWith("./") || importPath.startsWith("../")) {
                    var absolutePath = path.resolve(path.dirname(metadata.path), importPath);
                    return {
                        importPath: path.relative(targetDirectory, absolutePath),
                        exportedDeclaration: exportedDeclaration,
                    };
                }
                else {
                    return { importPath: importPath, exportedDeclaration: exportedDeclaration };
                }
            }
        }
    }
    return {
        importPath: null,
        exportedDeclaration: undefined,
    };
}
function findImportPathOrThrow(importName, targetDirectory, metadata) {
    var ret = findImportPath(importName, targetDirectory, metadata);
    if (!ret.importPath) {
        throw new Error("".concat(importName, " import not found in ").concat(metadata.path));
    }
    return ret;
}
function findEnumName(propertyName, metadata) {
    var tsProp = morphTsProperty(propertyName, metadata);
    return tsProp
        .getType()
        .getText(tsProp)
        .replace(/ ?\| ?undefined$/, "")
        .replace(/\[\]$/, "");
}
function findEnumImportPath(enumName, targetDirectory, metadata) {
    var _a;
    var tsSource = morphTsSource(metadata);
    if (tsSource.getEnum(enumName)) {
        //enum defined in same file as entity
        if (!((_a = tsSource.getEnum(enumName)) === null || _a === void 0 ? void 0 : _a.isExported())) {
            throw new Error("Enum ".concat(enumName, " is not exported in ").concat(metadata.path));
        }
        return path.relative(targetDirectory, metadata.path).replace(/\.ts$/, "");
    }
    else {
        //try to find import where enum is imported from
        var importPath = findImportPathOrThrow(enumName, targetDirectory, metadata).importPath;
        return importPath;
    }
}
function findValidatorImportPath(validatorName, generatorOptions, metadata) {
    var _a, _b;
    var tsSource = morphTsSource(metadata);
    //validator defined in same file as entity
    if (tsSource.getVariableDeclaration(validatorName) || tsSource.getFunction(validatorName)) {
        if (!(((_a = tsSource.getVariableDeclaration(validatorName)) === null || _a === void 0 ? void 0 : _a.isExported()) || ((_b = tsSource.getFunction(validatorName)) === null || _b === void 0 ? void 0 : _b.isExported()))) {
            throw new Error("Validator ".concat(validatorName, " is not exported in ").concat(metadata.path));
        }
        return path.relative("".concat(generatorOptions.targetDirectory, "/dto"), metadata.path).replace(/\.ts$/, "");
    }
    else {
        var importPath = findImportPathOrThrow(validatorName, generatorOptions.targetDirectory, metadata).importPath;
        return importPath;
    }
}
function findBlockName(propertyName, metadata) {
    var tsProp = morphTsProperty(propertyName, metadata);
    if (!tsProp)
        throw new Error("property not found");
    var rootBlockDecorator = tsProp.getDecorators().find(function (i) { return i.getName() == "RootBlock"; });
    if (!rootBlockDecorator)
        throw new Error("RootBlock decorator not found for property ".concat(propertyName));
    return rootBlockDecorator.getArguments()[0].getText();
}
function findBlockImportPath(blockName, targetDirectory, metadata) {
    var _a;
    var tsSource = morphTsSource(metadata);
    if (tsSource.getVariableDeclaration(blockName)) {
        //block defined in same file as entity
        if (!((_a = tsSource.getVariableDeclaration(blockName)) === null || _a === void 0 ? void 0 : _a.isExported())) {
            throw new Error("Block ".concat(blockName, " is not exported in ").concat(metadata.path));
        }
        return path.relative(targetDirectory, metadata.path).replace(/\.ts$/, "");
    }
    else {
        //try to find import where block is imported from
        var importPath = findImportPathOrThrow(blockName, targetDirectory, metadata).importPath;
        return importPath;
    }
}
function findInputClassImportPath(className, targetDirectory, metadata) {
    var _a;
    var tsSource = morphTsSource(metadata);
    var returnImportPath;
    var classDeclaration;
    if (tsSource.getClass(className)) {
        //block defined in same file as entity
        if (!((_a = tsSource.getClass(className)) === null || _a === void 0 ? void 0 : _a.isExported())) {
            throw new Error("Class ".concat(className, " is not exported in ").concat(metadata.path));
        }
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        classDeclaration = tsSource.getClass(className);
        returnImportPath = path.relative(targetDirectory, metadata.path).replace(/\.ts$/, "");
    }
    else {
        //try to find import where block is imported from
        var _b = findImportPathOrThrow(className, targetDirectory, metadata), importPath = _b.importPath, exportedDeclaration = _b.exportedDeclaration;
        if (!(exportedDeclaration instanceof ts_morph_1.ClassDeclaration)) {
            throw new Error("Exported declaration for ".concat(className, " is not a class"));
        }
        classDeclaration = exportedDeclaration;
        returnImportPath = importPath;
    }
    if (!classDeclaration.getDecorators().some(function (i) { return i.getName() == "InputType"; })) {
        throw new Error("Exported declaration for ".concat(className, " is not decorated with @InputType"));
    }
    return returnImportPath;
}
function getFieldDecoratorClassName(propertyName, metadata) {
    var definedDecorators = morphTsProperty(propertyName, metadata).getDecorators();
    var fieldDecorator = definedDecorators.find(function (decorator) { return decorator.getName() == "Field"; });
    if (!fieldDecorator) {
        return false;
    }
    var typeFnArg = fieldDecorator.getArguments()[0];
    if (!typeFnArg)
        throw new Error("".concat(propertyName, ": @Field is missing argument"));
    if (typeFnArg.getKind() != ts_morph_1.SyntaxKind.ArrowFunction)
        throw new Error("".concat(propertyName, ": @Field first argument must be an ArrowFunction"));
    var typeReturningArrowFunction = typeFnArg;
    var body = typeReturningArrowFunction.getBody();
    if (body.getKind() != ts_morph_1.SyntaxKind.Identifier)
        throw new Error("".concat(propertyName, ": @Field first argument must be an ArrowFunction returning an Identifier"));
    var identifier = body;
    var className = identifier.getText();
    return className;
}
