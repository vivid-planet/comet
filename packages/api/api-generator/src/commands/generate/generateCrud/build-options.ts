import { type CrudGeneratorOptions, getCrudSearchFieldsFromMetadata, hasCrudFieldFeature, SCOPED_ENTITY_METADATA_KEY } from "@comet/cms-api";
import { type EntityMetadata } from "@mikro-orm/core";

import { buildNameVariants } from "../utils/build-name-variants";
import { integerTypes } from "../utils/constants";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function buildFilterProps(metadata: EntityMetadata<any>) {
    const dedicatedResolverArgProps = buildDedicatedResolverArgProps(metadata);

    const crudFilterProps = metadata.props.filter(
        (prop) =>
            hasCrudFieldFeature(metadata.class, prop.name, "filter") &&
            !prop.name.startsWith("scope_") &&
            prop.name != "position" &&
            (!prop.embedded || hasCrudFieldFeature(metadata.class, prop.embedded[0], "filter")) && // the whole embeddable has filter disabled
            (prop.enum ||
                prop.type === "string" ||
                prop.type === "text" ||
                prop.type === "DecimalType" ||
                prop.type === "number" ||
                integerTypes.includes(prop.type) ||
                prop.type === "BooleanType" ||
                prop.type === "boolean" ||
                prop.type === "DateType" ||
                prop.type === "Date" ||
                prop.kind === "m:1" ||
                prop.kind === "1:m" ||
                prop.kind === "m:n" ||
                prop.type === "EnumArrayType" ||
                prop.type === "uuid") &&
            !dedicatedResolverArgProps.some((dedicatedResolverArgProp) => dedicatedResolverArgProp.name == prop.name),
    );
    return crudFilterProps;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function buildSortProps(metadata: EntityMetadata<any>) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    function directSortProps(metadata: EntityMetadata<any>) {
        return metadata.props
            .filter(
                (prop) =>
                    hasCrudFieldFeature(metadata.class, prop.name, "sort") &&
                    !prop.name.startsWith("scope_") &&
                    (!prop.embedded || hasCrudFieldFeature(metadata.class, prop.embedded[0], "sort")) && // the whole embeddable has sort disabled
                    (prop.type === "string" ||
                        prop.type === "text" ||
                        prop.type === "DecimalType" ||
                        prop.type === "number" ||
                        integerTypes.includes(prop.type) ||
                        prop.type === "BooleanType" ||
                        prop.type === "boolean" ||
                        prop.type === "DateType" ||
                        prop.type === "Date" ||
                        prop.kind === "m:1" ||
                        prop.type === "EnumArrayType" ||
                        prop.enum),
            )
            .map((prop) => {
                return prop.name;
            });
    }
    const crudSortProps = directSortProps(metadata);

    // add nested from relations, one level deep
    metadata.props.forEach((prop) => {
        if (hasCrudFieldFeature(metadata.class, prop.name, "sort")) {
            if ((prop.kind == "1:1" || prop.kind == "m:1") && prop.targetMeta) {
                crudSortProps.push(...directSortProps(prop.targetMeta).map((nestedProp) => `${prop.name}.${nestedProp}`));
                return true;
            }
        }
    });

    return crudSortProps;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function buildDedicatedResolverArgProps(metadata: EntityMetadata<any>) {
    return metadata.props.filter((prop) => {
        if (hasCrudFieldFeature(metadata.class, prop.name, "dedicatedResolverArg")) {
            if (prop.kind == "m:1") {
                return true;
            } else {
                console.warn(`${metadata.className} ${prop.name} can't use dedicatedResolverArg as it's not a m:1 relation`);
                return false;
            }
        }
        return false;
    });
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function buildOptions(metadata: EntityMetadata<any>, generatorOptions: CrudGeneratorOptions) {
    const { classNameSingular, classNamePlural, fileNameSingular, fileNamePlural } = buildNameVariants(metadata);

    const dedicatedResolverArgProps = buildDedicatedResolverArgProps(metadata);

    const crudSearchPropNames = getCrudSearchFieldsFromMetadata(metadata);
    const hasSearchArg = crudSearchPropNames.length > 0;

    const crudFilterProps = buildFilterProps(metadata);
    const hasFilterArg = crudFilterProps.length > 0;

    const crudSortProps = buildSortProps(metadata);
    const hasSortArg = crudSortProps.length > 0;

    const hasSlugProp = metadata.props.some((prop) => prop.name == "slug");

    const scopeProp = metadata.props.find((prop) => prop.name == "scope");
    if (scopeProp && !scopeProp.targetMeta) throw new Error("Scope prop has no targetMeta");

    const hasPositionProp = metadata.props.some((prop) => prop.name == "position");

    const positionGroupPropNames: string[] = hasPositionProp
        ? (generatorOptions.position?.groupByFields ?? [
              ...(scopeProp ? [scopeProp.name] : []), // if there is a scope prop it's effecting position-group, if not groupByFields should be used
          ])
        : [];
    const positionGroupProps = hasPositionProp ? metadata.props.filter((prop) => positionGroupPropNames.includes(prop.name)) : [];

    const scopedEntity = Reflect.getMetadata(SCOPED_ENTITY_METADATA_KEY, metadata.class);
    const skipScopeCheck = !scopeProp && !scopedEntity;

    const argsClassName = `${classNameSingular != classNamePlural ? classNamePlural : `${classNamePlural}List`}Args`;
    const argsFileName = `${fileNameSingular != fileNamePlural ? fileNamePlural : `${fileNameSingular}-list`}.args`;

    const blockProps = metadata.props.filter((prop) => {
        return hasCrudFieldFeature(metadata.class, prop.name, "input") && prop.type === "RootBlockType";
    });

    return {
        crudSearchPropNames,
        hasSearchArg,
        crudFilterProps,
        hasFilterArg,
        crudSortProps,
        hasSortArg,
        hasSlugProp,
        hasPositionProp,
        positionGroupProps,
        scopeProp,
        skipScopeCheck,
        argsClassName,
        argsFileName,
        blockProps,
        dedicatedResolverArgProps,
    };
}
