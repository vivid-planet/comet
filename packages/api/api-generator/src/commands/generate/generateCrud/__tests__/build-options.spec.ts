import { BaseEntity, defineConfig, Entity, ManyToOne, MikroORM, OneToOne, PrimaryKey, Property, Ref } from "@mikro-orm/postgresql";
import { LazyMetadataStorage } from "@nestjs/graphql/dist/schema-builder/storages/lazy-metadata.storage";
import { v4 as uuid } from "uuid";

import { testPermission } from "../../utils/test-helper";
import { buildSortProps } from "../build-options";
import { generateSortDto } from "../generate-crud";

@Entity()
export class TestEntity1 extends BaseEntity {
    @PrimaryKey({ type: "uuid" })
    id: string = uuid();

    @Property()
    foo: string;
}

@Entity()
export class TestEntity2 extends BaseEntity {
    @PrimaryKey({ type: "uuid" })
    id: string = uuid();

    @ManyToOne(() => TestEntityTarget, { nullable: true, index: true, ref: true })
    nested?: Ref<TestEntityTarget> = undefined;
}

@Entity()
export class TestEntity3 extends BaseEntity {
    @PrimaryKey({ type: "uuid" })
    id: string = uuid();

    @OneToOne(() => TestEntityTarget, { nullable: true, index: true, ref: true })
    nested?: Ref<TestEntityTarget> = undefined;
}

@Entity()
export class TestEntityTarget extends BaseEntity {
    @PrimaryKey({ type: "uuid" })
    id: string = uuid();

    @Property()
    foo: string;
}

describe("buildOptions", () => {
    describe("buildSortProps", () => {
        it("string field should be used as sortProp", async () => {
            LazyMetadataStorage.load();
            const orm = await MikroORM.init(
                defineConfig({
                    dbName: "test-db",
                    connect: false,
                    entities: [TestEntity1],
                }),
            );

            const sortProps = buildSortProps(orm.em.getMetadata().get("TestEntity1"));
            expect(sortProps).toEqual(["foo"]);

            await orm.close();
        });

        it("nested m:1 relation string field should be used as sortProp", async () => {
            LazyMetadataStorage.load();
            const orm = await MikroORM.init(
                defineConfig({
                    dbName: "test-db",
                    connect: false,
                    entities: [TestEntity2, TestEntityTarget],
                }),
            );

            const metadata = orm.em.getMetadata().get("TestEntity2");

            const sortProps = buildSortProps(metadata);
            expect(sortProps).toEqual(["nested", "nested.foo"]);

            const sortDto = generateSortDto({ generatorOptions: { targetDirectory: __dirname, requiredPermission: testPermission }, metadata });
            expect(sortDto).toMatchSnapshot();

            await orm.close();
        });

        it("nested 1:1 relation string field should be used as sortProp", async () => {
            LazyMetadataStorage.load();
            const orm = await MikroORM.init(
                defineConfig({
                    dbName: "test-db",
                    connect: false,
                    entities: [TestEntity3, TestEntityTarget],
                }),
            );

            const sortProps = buildSortProps(orm.em.getMetadata().get("TestEntity3"));
            expect(sortProps).toEqual(["nested.foo"]);

            await orm.close();
        });
    });
});
