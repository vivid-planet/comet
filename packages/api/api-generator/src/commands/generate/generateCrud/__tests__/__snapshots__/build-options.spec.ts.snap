// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildOptions buildSortProps nested m:1 relation string field should be used as sortProp 1`] = `
"import { SortDirection } from "@comet/cms-api";
    import { Field, InputType, registerEnumType } from "@nestjs/graphql";
    import { Type } from "class-transformer";
    import { IsEnum } from "class-validator";

    export enum TestEntity2SortField {
        nested = "nested",
nested_foo = "nested_foo",
    }
    registerEnumType(TestEntity2SortField, {
        name: "TestEntity2SortField",
    });
    
    @InputType()
    export class TestEntity2Sort {
        @Field(() => TestEntity2SortField)
        @IsEnum(TestEntity2SortField)
        field: TestEntity2SortField;
    
        @Field(() => SortDirection, { defaultValue: SortDirection.ASC })
        @IsEnum(SortDirection)
        direction: SortDirection = SortDirection.ASC;
    }
    "
`;
